Added the combobox sample, in the 30_Advanced section.

diogo-bastos opened the pull request:
This is a sample containing a drop down list combo box. I created this sample as I needed a combo similar to ones in the Chosen or JQuery libraries for my most recent project.

diogo-bastos commented:
I made the git commit with the wrong email. I added that email to my git account now.

diogo-bastos commented:
It seems the travis build failed with the error "The attribute 'option' may not appear in tag 'li'". However, the 'option' attribute is used for an 'li' tag in the amp-selector example: [](https://ampbyexample.com/components/amp-selector/). Is there something I am not doing correctly?

diogo-bastos commented:
That sounds like a good idea that would make the sample a lot more dynamic! I will implement this change this weekend. 

I would not use it for static options though, because having the options already in the DOM is more responsive than waiting for an additional RTT when the user opens the list.

sebastianbenz commented:
Hm - curious how much of a difference it makes.

diogo-bastos commented:
Changed the client side code so that the dropdown options are loaded via amp-list with a mustache template. Sebastian, was this what you had in mind? I still need to add the server side file of this example though.

sebastianbenz commented on src/30_Advanced/ComboBox.html:
```diff
@@ -0,0 +1,177 @@
+                        id="autosuggest-list"
+                        hidden>
+                <template type="amp-mustache">
+                    <amp-selector keyboard-select-mode="focus"
```
this will have to go outside the amp-list.

sebastianbenz commented:
Almost. My idea was to still have the options json embedded in the page in an amp-state element (this avoids a server-request):

```
<amp-state id="options">
  <script type="application/json">
    {
       items: [
          {"title": "..., ...},
          {"title": "..., ...},
       ]
    }
  </script>
</amp-state>
```

Then you can bind the amp-list `src` to the options object:

```
   <amp-list [src]="options" ...>
```

And update the options object when the input field is focused:

```
<input ... on="tap:AMP.setState({options: options}), ...">
```

You might have to experiment with how to best trigger the amp-list refresh. 

Not sure this works though. ;-)

sebastianbenz commented:
Actually, it would be nice to have two samples, one demonstrating the local approach and one with a remote JSON list. (the remote one probably first as the markup is simpler). You can put a JSON file into src/json/food-options.json and then use it like this: `<amp-list src="/json/food-options.json" ...>`.

diogo-bastos commented:
Alright, I will to try to finish the remote sample tonight then. As for the local sample, should I make it in the same html file or in a different one?

sebastianbenz commented:
Great! Same file would be good I think.
On Tue, 7 Nov 2017 at 15:13, Diogo Bastos <notifications@github.com> wrote:

> Alright, I will to try to finish the remote sample tonight then. As for
> the local sample, should I make it in the same html file or in a different
> one?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amp-by-example/pull/1001#issuecomment-342513055>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXOODzBrXZa5VvYdZeXYfBsMN0cQjnsks5s0HOZgaJpZM4QO00E>
> .
>

diogo-bastos commented:
Added the json file and fixed the template for the remote sample.

As for the local sample, I tried adding the amp-list "src" attribute as the id of an amp-state ("options" in this case) but the amp-list component just tried to make a GET request for some remote endpoint "options", instead of binding to the amp-state. Is there something I'm missing?

sebastianbenz commented on src/30_Advanced/ComboBox.html:
```diff
@@ -0,0 +1,181 @@
+  <!-- ## Basic Usage -->
+  <!--
+  This example includes two inputs. One is the input in which the user writes the text and the other is a hidden input, 
+  where the id value of the selected option is stored and could be submited in a form.
```
s/submited/submitted

diogo-bastos replied:
Done!

sebastianbenz commented on src/30_Advanced/ComboBox.html:
```diff
@@ -0,0 +1,181 @@
+  <!-- ## Setup -->
+  <!--
+  The amp-selector component is used for the dropdown list options. The amp-bind component is used for the filtering of the dropdown options
+  according to the user's input. The fontawesome css icons are imported for the arrow icon.
```
Can you document each import separately? Otherwise they won't be visible in the doc view of the sample.

diogo-bastos replied:
Done!

sebastianbenz commented on src/30_Advanced/ComboBox.html:
```diff
@@ -0,0 +1,181 @@
+  <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>
+  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
+  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.1.js"></script>
+  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/3.2.1/css/font-awesome.min.css">
```
Would you mind not using font-awesome here? I find it a bit overkill for a single icon (as it quite impacts loading performance). We usually use material icons, e.g. [this](https://material.io/icons/?utm_source=google&utm_medium=cpc&utm_campaign=1001467%20%7C%20Material.IO%20%7C%20Global%20%7C%20en%20%7C%20Hybrid%20%7C%20Text%20%7C%20BKWS&utm_term=%7Bkeyword%7D#ic_keyboard_arrow_down) and [this](https://material.io/icons/?utm_source=google&utm_medium=cpc&utm_campaign=1001467%20%7C%20Material.IO%20%7C%20Global%20%7C%20en%20%7C%20Hybrid%20%7C%20Text%20%7C%20BKWS&utm_term=%7Bkeyword%7D#ic_keyboard_arrow_up) one. You could either copy them to `src/img/*` or inline them via css (what I'd recommend doing).

diogo-bastos replied:
Alright, I don't mind. Removed the font-awesome import and added material icons instead.

sebastianbenz commented on src/30_Advanced/ComboBox.html:
```diff
@@ -0,0 +1,181 @@
+            padding: 5px 8px;
+            font-size: 1em;
+            background-color: #fff;
+            font-family: Arial,"Helvetica Neue",Helvetica,sans-serif;
```
Is there a particular reason why you're using a custom font here?

diogo-bastos replied:
Not really, it was just a mistake. I removed this in the latest commit.

sebastianbenz commented on src/30_Advanced/ComboBox.html:
```diff
@@ -0,0 +1,181 @@
+        .select-option:focus {
+            background: #fff;
+        }
+
```
nit: remove empty line

diogo-bastos replied:
Done!

sebastianbenz commented on src/30_Advanced/ComboBox.html:
```diff
@@ -0,0 +1,181 @@
+  This example includes two inputs. One is the input in which the user writes the text and the other is a hidden input, 
+  where the id value of the selected option is stored and could be submited in a form.
+
+  The dropdown list's options are rendered server side, instead of using amp-list to maximize responsiveness. 
```
I think this no longer applies.

diogo-bastos replied:
Removed.

sebastianbenz commented on src/30_Advanced/ComboBox.html:
```diff
@@ -0,0 +1,181 @@
+  where the id value of the selected option is stored and could be submited in a form.
+
+  The dropdown list's options are rendered server side, instead of using amp-list to maximize responsiveness. 
+  The 'option' attribute of the amp-selector's options, contain the id of the element, as well as its title, seperated by a comma.
```
s/'option'/`option`

nice trick with storing title and id in the option 👍

diogo-bastos replied:
Done!

sebastianbenz commented:
Regarding amp-list binding. Yeah, that's expected. Currently, you can't omit the src attribute of an amp-list, so it will always try to make an request. Maybe it's best to leave the server-side rendered until there is a better approach. WDYT?

diogo-bastos commented:
I think so too. I'll come back to this example if I remember some way to do this locally or if another option becomes available in the amp standard.

diogo-bastos commented:
Thank you very much for your patience and guidance, Sebastian! See you soon!
